src/Compiler/RefC/RefC.idr

```diff
@@ -363,76 +358,89 @@ emit fc line = do
         (Yes _) => flip snoc (lJust indentedLine maxLineLengthForComment ' ' ++ " " ++ comment)
         (No _)  => flip appendR [indentedLine, (lJust ""   maxLineLengthForComment ' ' ++ " " ++ comment)]
 
+addHeader : {auto h : Ref HeaderFiles (SortedSet String)}
+         -> String
+         -> Core ()
+addHeader = update HeaderFiles . insert
+
+applyFunctionToVars : {auto oft : Ref OutfileText Output}
+                    -> {auto il : Ref IndentLevel Nat}
+                    -> String
+                    -> List String
+                    -> Core $ ()
+applyFunctionToVars fun vars = traverse_ (\v => emit EmptyFC $ fun ++ "(" ++ v ++ ");" ) vars
 
-freeTmpVars : {auto t : Ref TemporaryVariableTracker (List (List String))}
-           -> {auto oft : Ref OutfileText Output}
+removeVars : {auto oft : Ref OutfileText Output}
            -> {auto il : Ref IndentLevel Nat}
+           -> List String
            -> Core $ ()
-freeTmpVars = do
-    lists <- get TemporaryVariableTracker
-    case lists of
-        (vars :: varss) => do
-            traverse_ (\v => emit EmptyFC $ "removeReference(" ++ v ++ ");" ) vars
-            put TemporaryVariableTracker varss
-        [] => pure ()
+removeVars = applyFunctionToVars "removeReference"
 
+dupVars : {auto oft : Ref OutfileText Output}
+           -> {auto il : Ref IndentLevel Nat}
+           -> List String
+           -> Core $ ()
+dupVars = applyFunctionToVars "newReference"
 
-addHeader : {auto h : Ref HeaderFiles (SortedSet String)}
-         -> String
-         -> Core ()
-addHeader = update HeaderFiles . insert
+removeReuseConstructors : {auto oft : Ref OutfileText Output}
+                        -> {auto il : Ref IndentLevel Nat}
+                        -> List String
+                        -> Core $ ()
+removeReuseConstructors = applyFunctionToVars "removeReuseConstructor"
 
+avarToC : Env -> AVar -> Core String
+avarToC env var = do
+    if contains var env.borrowed then pure $ "newReference(" ++ varName var ++ ")"
+        else if contains var env.owned then pure $ varName var
+        else coreFail $ InternalError $ "INTERNAL ERROR: variable " ++ varName var ++ " is not owned and borrowed "
```

Comment:
If every single variable is either borrowed or owned, why are we maintaining two sets?
Surely one is enough and the other is literally everything else?

```suggestion
        else coreFail $ InternalError $ "INTERNAL ERROR: variable " ++ varName var ++ " is neither owned nor borrowed "
```